rust/headless-client/src/ipc_service.rs

```diff
@@ -192,42 +211,72 @@ impl Handler {
         })
     }
 
-    // Infallible so that we only give up on an IPC client explicitly
-    async fn run(&mut self) {
+    /// Run the event loop to communicate with an IPC client.
+    ///
+    /// If the IPC service needs to terminate, we catch that from `signals` and send
+    /// the client a hint to shut itself down gracefully.
+    ///
+    /// The return type is infallible so that we only give up on an IPC client explicitly
+    async fn run(&mut self, signals: &mut signals::Terminate) -> HandlerOutcome {
         loop {
-            let event = {
-                // This borrows `self` so we must drop it before handling the `Event`.
-                let cb = pin!(self.cb_rx.recv());
-                match future::select(self.ipc_rx.next(), cb).await {
-                    future::Either::Left((Some(Ok(x)), _)) => Event::Ipc(x),
-                    future::Either::Left((Some(Err(error)), _)) => {
-                        tracing::error!(?error, "Error while deserializing IPC message");
-                        continue;
+            let event = poll_fn(|cx| {
```

Comment:
It might be worth extracting a dedicated poll function for this to hide the details from the flow of the loop.